slub/dfg-viewer
# Migrate to Kitodo.Presentation 4

This PR attempts to finish the migration to Kitodo.Presentation 4 and Extbase/Fluid, and does some maintenance and fixes along the way. When merged, this resolves #201 and #205.

The migration relies on the following PRs to other repos:
- https://github.com/kitodo/kitodo-presentation/pull/844
  (Needed to replace the custom calendar plugin)
- https://github.com/slub/slub_digitalcollections/pull/47

So to test this:
- Checkout the Kitodo and Digital Collections PRs. (There's also https://github.com/dvoracek-slub/kitodo-presentation/tree/dev-dfgviewer.)
- https://github.com/slub/ddev-dfgviewer/pull/2 is an update to the default DDEV system.
- https://github.com/dvoracek-slub/ddev-dfgviewer-dist/ may be used to easily test a clean install.

Questions:
- `language.typoscript` has a language condition that looks obsolete to me (when using website language and site configuration). Is this right?
- How should localization of metadata labels and structures be handled?
  In principle, they would work if the following is done:
  - Create a translation of the viewer page.
    - If we want to include it in the initialization data dump, we could probably add it manually. (The impexp extension doesn't export page translations as of 10.4, see https://forge.typo3.org/issues/93470.)
    - Alternatively, we can just add a remark in the documentation to do this after install if wanted.
  - Remove the `0` suffixes of column `index_name` in tables `tx_dlf_metadata` and `tx_dlf_structures`.
    - I tried to change `data.xml` to avoid these suffixes on fresh install, but couldn't find a way so far. Even when reordering the entries, it looks like TYPO3's `DataHandler::getUnique()` won't detect the translated records as such because the `l18n_parent` doesn't yet have an ID.
    - What does seem to work is an SQL query like the following. This could possibly be integrated into the Kitodo backend module or documented. 
      ```sql
      UPDATE `tx_dlf_metadata` a
      INNER JOIN `tx_dlf_metadata` b
      ON (a.l18n_parent = b.uid OR a.uid = b.l18n_parent)
      SET a.index_name = b.index_name
      WHERE a.index_name LIKE CONCAT(b.index_name, '0');

      UPDATE `tx_dlf_structures` a
      INNER JOIN `tx_dlf_structures` b
      ON (a.l18n_parent = b.uid OR a.uid = b.l18n_parent)
      SET a.index_name = b.index_name
      WHERE a.index_name LIKE CONCAT(b.index_name, '0');
      ```
